### ISSUE TITLE
indexeddb adapter: Database has a global failure

### ISSUE THREAD

**Issue by hermitdemschoenenleben** [pouchdb/pouchdb#8329]
I maintain a mobile app built using capacitor and pouchdb. By chance, I found out that sometimes "Database has a global failure" appears console.error'd (although no proper exception is thrown in this case). The code line that logs these errors is this one: 
https://github.com/pouchdb/pouchdb/blob/master/packages/node_modules/pouchdb-adapter-indexeddb/src/setup.js#L145

The detailed error messages are:

* UnknownError: Attempt to open a cursor in database without an in-progress transaction
* UnknownError: Attempt to iterate a cursor that doesn't exist
* UnknownError: Connection to Indexed Database server lost. Refresh the page to try again
* UnknownError: An internal error was encountered in the Indexed Database server

These errors only happen on ios, after the app is in background. After some researching I think that they most likely happen because ios locks the file system when an app is backgrounded, causing database operations to fail (cf. https://github.com/firebase/firebase-js-sdk/issues/2581 , https://github.com/firebase/firebase-js-sdk/issues/2232 . These issues are for firebase which also uses indexeddb). In other words, the culprit isn't pouchdb, but ios itself.

What I don't know now is what to do if these errors occur. Is "Database has a global failure" a recoverable error? Can pouchdb continue operating after something like this happened or is it a fatal error and the only thing I can do is to recreate pouchdb / reload the page?

In case that I do have to reload the page if this error happens: how can I catch it? So far, I only saw it on console.error. Trying to listen to database events `error`, `cancel`, `cancelled`, `close`, `closed` did not work.

**Comment by dheimoz** [comment]
Hey @hermitdemschoenenleben do you have a repo that I can test?
Which iOS version you are using?

**Comment by hermitdemschoenenleben** [comment]
Hi @dheimoz , thanks for your answer!

I have not yet been able to reproduce the issue on my device, but I'm working on it. I fear that it's going to be hard to reproduce as the error seems to happen if the application has been in background for quite some time. It does not happen all the time, though, but sporadically. Guessing from the error logs, I'd say that it happens for roughly 1% of my users per day. Affected ios versions seem to be 14.4, 14.5 and 14.6. I'll try next week to reproduce the issue with a testing repo though.

Until then: is it possible to guess how grave the error is? Will pouchdb be able to continue after it or should I reload the page if it happens in order to be on the safe side? The code that logs the error (
https://github.com/pouchdb/pouchdb/blob/master/packages/node_modules/pouchdb-adapter-indexeddb/src/setup.js#L147 ) executed `idb.close();` afterwards. Is there a mechanism in place to re-open idb again?

**Comment by hermitdemschoenenleben** [comment]
ok, I've tried to reproduce the errors but haven't succeded yet :( The logs I have indicate that the errors mainly happen if the app is in background. So I tried to write to the database in a loop, but this did not trigger the error.

As a workaround I'll just show an alert message and reload the page if "Database has a global failure" occurs because I currently have no idea how severe the error is or whether pouchdb can recover from it

**Comment by hermitdemschoenenleben** [comment]
I discovered another error in the logs, this time happening on android (9 / 10 / 11):

`Database has a global failure TimeoutError: Transaction timed out due to inactivity.`

This one apparently happens if the app runs for a long time without any writes to the database.

Does anybody know whether this kind of message apparently related to transactions being cancelled / closed are something that pouchdb already takes care of by retrying the operation?

**Comment by hermitdemschoenenleben** [comment]
Does anybody know whether this kind of message apparently related to transactions being cancelled / closed are something that pouchdb already takes care of by retrying the operation?

**Comment by nilsd** [comment]
@hermitdemschoenenleben Hi! Did you ever come to any conclusion regarding this issue? We're having the same problem in production but struggles to reproduce on our own devices.

**Comment by hermitdemschoenenleben** [comment]
@nilsd I didn't really come to a conclusion to how severe the error is. As I feared that it could affect data integrity or keep the database in an unresponsive state, I implemented two hacky workarounds:

# Lock database when the app is in background

In `pouchdb-adapter-indexeddb`, I patched `$`, `$p` and `$t` (these functions return database handles) such that they wait until the app is in foreground before they complete. This effectively pauses Pouchdb operation when the app is in background mode. See my hacky implementation of `waitForForeground` here: https://github.com/hermitdemschoenenleben/pouchdb/blob/blindmate/packages/node_modules/pouchdb-adapter-indexeddb/src/index.js#L39
This workaround reduced the number of occurrences by a large amount.

I then set `window.waitForForegroundPromise` if Capacitor tells me the app is in background mode:
```javascript

  /** we saw in production that database calls when the app is in paused state
   * are risky: sometimes they fail, putting the database in an undefined state.
   * Therefore, we implemented a hack in pouchdb that pauses all database operation
   * if `window.waitForForegroundPromise` doesn't emit a value.
   */
  waitForForegroundPromiseResolve: Function = undefined;
  lockDatabaseOperationsInBackground$ = createEffect(() => mergeObs(
    this.platform.pause.pipe(map(() => true)),
    this.platform.resume.pipe(map(() => false)),
    this.actions$.pipe(ofType(run.core.pauseDatabaseOperation), map(action => action.pause))
  ).pipe(
    switchMap(pausing => of(pausing).pipe(
      // when the app enters the paused state, we add a short delay before
      // actually locking the database because we may still want to write some
      // data
      delayWhen(pausing => pausing ? timer(1000) : timer(0)),
      map(pausing => {
        console.log('lock database operations: ', pausing)
        if (pausing) {
          // the app enters the paused state: create windowwaitForForegroundPromise
          // that is going to block any further database operation. This promise
          // will be resolved once the app resumes.
          if (!(window as any).waitForForegroundPromise) {
            (window as any).waitForForegroundPromise = new Promise((resolve) => {
              this.waitForForegroundPromiseResolve = resolve;
            });
          } else {
            console.log('database was already locked');
          }

        } else {
          // resuming app: resolve the promise such that database operation may
          // continue
          if (this.waitForForegroundPromiseResolve) this.waitForForegroundPromiseResolve();
          (window as any).waitForForegroundPromise = undefined;
        }
      })
    ))
  ), { dispatch: false });

```

# Reload application if global failure happens
Even with the workaround, I sometimes see "Database has a global failure". To be on the safe side, I reload the app in this case. As this mostly happens when the app is / was in background, users don't notice it (I also show a splash screen at the same time).
To catch the "global failure", I simply patched a global callback: https://github.com/hermitdemschoenenleben/pouchdb/blob/blindmate/packages/node_modules/pouchdb-adapter-indexeddb/src/setup.js#L148

To sum up: I'm not super happy with the workarounds, but they are "good enough" for me at the moment. Having my app in production with a few thousand daily users, I don't see major problems with the database with these fixes.

**Comment by jcoglan** [comment]
I expect that for most applications it would be surprising for PouchDB calls to block on the page being foregrounded, it would probably be more useful for applications to be able to detect these failures to handle them if desired. The main interaction points that seem to be available for us to detect the problem are:

- Calling a PouchDB method will call [`indexedDB.open()`](https://github.com/pouchdb/pouchdb/blob/3fe71b036a37f3cf983f47937bd9ccd0788fead6/packages/node_modules/pouchdb-adapter-indexeddb/src/setup.js#L108-L109) if this is the first time the DB is accessed, or if `onversionchange` has triggered, probably due to a new index definition. Otherwise the existing DB handle is reused.
- Many methods call [`transaction()`](https://github.com/pouchdb/pouchdb/blob/3fe71b036a37f3cf983f47937bd9ccd0788fead6/packages/node_modules/pouchdb-adapter-indexeddb/src/index.js#L81) in order to access the DB data. If this throws an exception, we may be able to catch it and distinguish these non-recoverable failures from other transient errors.
- The [`onabort` event](https://github.com/pouchdb/pouchdb/blob/3fe71b036a37f3cf983f47937bd9ccd0788fead6/packages/node_modules/pouchdb-adapter-indexeddb/src/setup.js#L144-L148) causes the error to be logged and the DB closed, dropping the cached DB handle so we'll need to call `open()` again the next time we need the DB.

It's not clear if any/all these failures bubble up through the public PouchDB API where an application could handle them. It may also be useful to emit an event on the `PouchDB` instance, or the constructor, to alert the application to these failures.

Can someone experiencing these failures check the above interaction points to see what error information we're able to gather, and whether it's possible to distinguish these non-recoverable failures from others?

**Comment by nilsd** [comment]
@hermitdemschoenenleben @jcoglan Thank you both for your helpful replies. We are running the sqlite adapter for now but we'll play around with your suggestions once we're stabilized. What I do know, though, is that our users are constantly switching between foreground/background since they are using turn-by-turn navigation in addition to our application. Reloading the app or re-initializing the database on resume will be the first thing we try when we return to the indexeddb adapter.

**Comment by nskoro** [comment]
having this same issue
